### ISSUE TITLE
Unsoundeness in SendRc

### ISSUE THREAD

**Issue by Skepfyr** [hniksic/sendable#1]
Hi! I believe I have spotted some unsoundness in the SendRc implementation, or more specifically [this line of `PostSend`](https://github.com/hniksic/sendable/blob/master/src/send_rc.rs#L378).

This is the reproducer I came up with, it (usually) doesn't actually behave unsoundly but it does show that there is a reference to the same RefCell on 2 different threads.
```rust
use std::{
    cell::RefCell,
    sync::{Arc, Barrier},
};

use sendable::SendRc;

fn main() {
    let mut a = SendRc::new(RefCell::new(1));
    let mut b = SendRc::clone(&a);

    let mut pre_send = SendRc::pre_send();
    pre_send.disable(&mut a);
    pre_send.disable(&mut b);
    let mut post_send = pre_send.ready();

    post_send.enable(&mut a);
    let a_inner = &*a;
    let a_barrier = Arc::new(Barrier::new(2));
    let b_barrier = Arc::clone(&a_barrier);
    std::thread::spawn(move || {
        post_send.enable(&mut b);
        drop(post_send);
        let b_inner = &*b;
        b_barrier.wait();
        println!("{}", b_inner.borrow_mut())
    });
    a_barrier.wait();
    println!("{}", a_inner.borrow_mut());
}
```

The idea is to exploit the line I linked above by re-enabling one copy of the `SendRc` on the original thread, before sending the other copy to another thread.

More generally, I think the idea behind the library *is* sound, and I didn't spot any other unsoundness but I'm not very confident in that.

**Comment by hniksic** [comment]
Thanks for the review and the example! I can confirm the unsoundness. The [fix](https://github.com/hniksic/sendable/commit/a65b24d3e954ade488604c165aa9edada5da0e2b) is to pin allocations to thread 0 (which is never assigned by `current_thread()`) in `PostSend::ready()`, and to check for that in `enable()`.

With the fix your example consistently panics in `PostSend::enable()` in the other thread, and hangs in the barrier in the main thread.

I additionally [disallowed](https://github.com/hniksic/sendable/commit/0b3a03d60536ba782d84001b7a7a9328e9977bf5) calling `PostSend::enable()` from different threads, even if that would be sound, as `enable()` checks for the pin for a particular allocation. But if one needs to migrate to multiple threads, they should use multiple `PostSend`s. Removing that freedom makes the code easier to reason about.

**Comment by hniksic** [comment]
Fix in 0.3.0

**Comment by hniksic** [comment]
@Skepfyr You might want to take a look at the v0.4.0 release, which significantly simplifies `PostSend`.
